File: StopWatch_oc/StopWatch_oc/StopWatchData.m
//
//  StopWatchData.m
//  StopWatch_oc
//
//  Created by ynfMac on 2020/5/25.
//  Copyright © 2020 ynfMac. All rights reserved.
//

#import "StopWatchData.h"

@interface StopWatchData()

@property (nonatomic,strong,readwrite) NSArray<NSNumber *> * times;

@end

@implementation StopWatchData
{
    NSMutableArray<NSNumber *> *_times;
}

#pragma mark initial
- (instancetype)init{
    if (self = [super init]) {
        [self reset];
    }
    
    return self;
}

#pragma mark method
- (void)reset{
    _isReset = YES;
    _times = [NSMutableArray array];
}

- (void)beginingNewTimeWithCompletion:( void (^)(void))completion{
    [_times insertObject:@0 atIndex:0];
    _isReset = NO;
    
    if (completion) {
        completion();
    }
}

- (void)timing:(NSTimeInterval)time{
    [_times replaceObjectAtIndex:0 withObject:@(time)];
}


#pragma mark getter
- (NSUInteger)maxTime{
    if (_times.count <= 2) {
        return 0;
    }
    
    int maxIndex = 1;
    for (int i = 2; i < _times.count; i++) {
        if ([_times[i] intValue] > [_times[maxIndex] intValue]) {
            maxIndex = i;
        }
    }
    return maxIndex;
}

- (NSUInteger)minTime{
    if (_times.count<= 2) {
        return 0;
    }
    
    int minIndex = 1;
    for (int i = 2; i < _times.count; i++) {
        if ([_times[i] intValue] < [_times[minIndex] intValue]) {
            minIndex = i;
        }
    }
    return minIndex;
}

- (NSArray<NSNumber *> *)allTimes{
    return _times;
}



@end

File: stopWatch_swift/stopWatch_swift/stopWatchData.swift
//
//  stopWatchData.swift
//  stopWatch_swift
//
//  Created by ynfMac on 2020/5/25.
//  Copyright © 2020 ynfMac. All rights reserved.
//

import Foundation

class stopWatchData {
    var isReset:Bool = true
    var times:[Int] = []
    
    var maxTime:Int? {
        get {
            if times.count <= 2 {
                return nil;
            }
            
            var maxIndex = 1
            
            for index in 2..<times.count {
                if(times[index] > times[maxIndex]){
                    maxIndex = index
                }
            }
            return maxIndex
        }
    }
    
    var minTime:Int? {
        get {
            if times.count <= 2 {
                return nil;
            }
            
            var minIndex = 1
            
            for index in 2..<times.count {
                if(times[index] < times[minIndex]){
                    minIndex = index
                }
            }
            return minIndex
        }
    }
    
    init() {
        reset()
    }
    
    func reset() {
        isReset = true
        times = []
    }
    
    
    func timing(time:Int) {
        times[0] = time
    }
    
    func beginingNewTime(comletion:(() ->())? = nil){
        times.insert(0, at: 0)
        isReset = false
        
        if let comletion = comletion {
            comletion()
        }
    }
    
    
}
